### PR TITLE
Support to Add, Remove and List ACLs (Fixes #6)

### LINKED ISSUE #6: Support to Add, Remove and List ACLs
### What problem are you facing? 

A fully feature complete provider-kafka needs to enable users to add, remove, and list ACLs just like a user would via the authorizer CLI.

### How could Crossplane help solve your problem?

An ACL controller in provider-kafka providing CRUD operations for ACLs could replicate the functionality of the authorizer CLI.

### PR DESCRIPTION
**Pull request by jograca**
### Description of your changes

This Pull Request adds feature complete ACL support to Add, Remove and List ACLs.
Fixes #6 

I have:

- [ x ] Read and followed Crossplane's [contribution process].
- [ x ] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Automated tests are forthcoming

1) Created an ACL by applying examples/acl/acl.yaml
2) Verified by running `kcl admin acl describe --type any --pattern match --op any --perm any`
3) Updated ACL by changing a Principle in examples/acl/acl.yml
4) Applied the ACL and noticed the Update threw an error
3) Deleted ACL by deleting examples/acl/acl.yaml
4) Verified by same process as bullet 2

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: internal/clients/kafka/acl/acl.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+type Topic struct {
+	Name              string
+	ReplicationFactor int16
+	Partitions        int32
+	ID                string
+	Config            map[string]*string
+}
+
+// List lists all the ACLs in Kafka
+func List(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) (*AccessControlList, error) {
```
[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:38]: Hmm, checking the terraform provider, it looks like it does not support update for ACLs, e.g. there is no update method implemented: https://github.com/Mongey/terraform-provider-kafka/blob/master/kafka/resource_kafka_acl.go

File: internal/clients/kafka/acl/acl.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+// AccessControlList is a holistic representation of a Kafka ACL with configurable
+// fields
+type AccessControlList struct {
+	Name                            string
+	ResourceType                    string
+	AccessControlListPrinciple      string
+	AccessControlListHost           string
+	AccessControlListOperation      string
+	AccessControlListPermissionType string
```
Comment (turkenh) [inline internal/clients/kafka/acl/acl.go:23]:
```suggestion
	Principle      string
	Host           string
	Operation      string
	PermissionType string
```

File: apis/acl/v1alpha1/acl_types.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// AccessControlListParameters are the configurable fields of a AccessControlList.
+type AccessControlListParameters struct {
+	// +kubebuilder:validation:Enum=Unknown;Any;Topic;Group;Cluster;TransactionalID
+	ResourceType               string `json:"resourceType"`
+	AccessControlListPrinciple string `json:"accessControlListPrinciple"`
+	AccessControlListHost      string `json:"accessControlListHost"`
+	// +kubebuilder:validation:Enum=Unknown;Any;All;Read;Write;Create;Delete;Alter;Describe;ClusterAction;DescribeConfigs;AlterConfigs;IdempotentWrite
+	AccessControlListOperation string `json:"accessControlListOperation"`
+	// +kubebuilder:validation:Enum=Unknown;Any;Allow;Deny
+	AccessControlListPermissionType string `json:"accessControlListPermissionType"`
```
Comment (turkenh) [inline apis/acl/v1alpha1/acl_types.go:37]:
```suggestion
	Principle string `json:"principle"`
	Host      string `json:"host"`
	// +kubebuilder:validation:Enum=Unknown;Any;All;Read;Write;Create;Delete;Alter;Describe;ClusterAction;DescribeConfigs;AlterConfigs;IdempotentWrite
	Operation string `json:"operation"`
	// +kubebuilder:validation:Enum=Unknown;Any;Allow;Deny
	PermissionType string `json:"permissionType"`
```

File: internal/clients/kafka/acl/acl.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	acl.AccessControlListPermissionType = accessControlList.AccessControlListPermissionType
+	acl.ResourcePatternTypeFilter = accessControlList.ResourcePatternTypeFilter
+
+	return &acl, nil
+}
+
+// Create creates an ACL from the Kafka side
+func Create(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) error {
+
+	o, _ := kmsg.ParseACLOperation(strings.ToLower(accessControlList.AccessControlListOperation))
```
Comment (turkenh) [inline internal/clients/kafka/acl/acl.go:69]: error needs to be handled

File: internal/clients/kafka/acl/acl.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	ao := []kadm.ACLOperation{o}
+
+	rpt, _ := kmsg.ParseACLResourcePatternType(strings.ToLower(accessControlList.ResourcePatternTypeFilter))
```
Comment (turkenh) [inline internal/clients/kafka/acl/acl.go:72]: error needs to be handled

File: internal/clients/kafka/acl/acl.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+		a := resp[0].Principal
+		if len(a) == 0 {
+			return errors.New("no create response for acl")
+		}
+
+		return nil
+
+	}
+
+	return nil
```
Comment (turkenh) [inline internal/clients/kafka/acl/acl.go:125]: Could be refactored as:

```suggestion
	b := kadm.ACLBuilder{}
	ab := b.Allow(accessControlList.AccessControlListPrinciple).AllowHosts(accessControlList.AccessControlListHost).Operations(ao[0]).ResourcePatternType(rpt)

	switch accessControlList.ResourceType {
	case "Topic":
		ab = ab.Topics(accessControlList.Name)
	case "Group":
		ab = ab.Groups(accessControlList.Name)
	case "TransactionalID":
		ab = ab.TransactionalIDs(accessControlList.Name)
	}

	resp, err := cl.CreateACLs(ctx, ab)
	if err != nil {
		return err
	}

	a := resp[0].Principal
	if len(a) == 0 {
		return errors.New("no create response for acl")
	}

	return nil
```

File: internal/clients/kafka/acl/acl.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// Delete creates an ACL from the Kafka side
+func Delete(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) error {
+
+	o, _ := kmsg.ParseACLOperation(strings.ToLower(accessControlList.AccessControlListOperation))
+	ao := []kadm.ACLOperation{o}
+
+	rpt, _ := kmsg.ParseACLResourcePatternType(strings.ToLower(accessControlList.ResourcePatternTypeFilter))
```
Comment (turkenh) [inline internal/clients/kafka/acl/acl.go:134]: Could be refactored as a separate function.

File: internal/controller/acl/acl.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+	log         logging.Logger
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.AccessControlList)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAccessControlList)
+	}
+
+	ae, _ := acl.List(ctx, c.kafkaClient, acl.Generate(meta.GetExternalName(cr), &cr.Spec.ForProvider))
```
Comment (turkenh) [inline internal/controller/acl/acl.go:138]: please handle the error.

File: internal/clients/kafka/acl/acl.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:28]: Ideally, our resource should reject any updates after creation, simply returning an error. However, this is [not possible](https://github.com/crossplane/crossplane/issues/727) today but hopefully will be, with an admission [webhook](https://github.com/crossplane/crossplane/issues/1152) in the near future.

There are other resources in other providers that do not support updates. However, we have another challenge here that is the lack of a unique identifier. 

> We could have ResourceUpToDate set to false, which would send a user to Create when they update any configs.

Wouldn't this still cause the creation of a new resource with updated config and losing track of the previous one as above 🤔

I am wondering if we can leverage `external-name` annotation here. Building a _unique_ identifier by concatenating all the fields (similar to what terraform provider uses for [import](https://github.com/Mongey/terraform-provider-kafka/pull/152#issuecomment-728120534)) and at the beginning of observe, compare it with the new spec and reject going further with a proper error like `updates not supported` if it is different.

This would require disabling [NameAsExternalName](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/api.go#L46) initializer like [this](https://github.com/crossplane/provider-aws/blob/b3f7c2b86595f053bf9295cd05dfac5a71d1535a/pkg/controller/servicediscovery/httpnamespace/setup.go#L60) and setting it once after create.

[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:28]: We've been thinking about this, and don't feel like we have a good solution.  I'm gonna @twmb to see if he has thoughts on whether this is feasible or not.  @schaubce @rscibelli too.

**Problem:**

Present state: One managed resource, say `marshmallory`, to all ACLs named `marshmallory`.  Because K8s doesn't have a unique ID, create runs, the first `marshmallory` gets orphaned, and the latest yaml is applied.  

Desired state: One managed resource, `marshmallory`, maps to a unique Kafka ACL.

**Three possible solutions:**

Caveats: Can't update, so it will be a create and delete only action. 

1. Try the UUID on the k8s side.

Example:
Managed resources:
1. `name: marshmallory-abcd`
2. `name: marshmallory-defg`

Kafka side resources:
1. kafname: `marshmallory`
2. kafname: `marshmallory`

So kafname 1 and 2 appear the same, but have different underlying configurations such as a different user name, or different permissions.

In order to verify that we have the *right* ACL, it has to be described given ALL of the properties.

Pros: 
-Seems like it may work
Cons: 
- I have a feeling this will be a performance issue on large clusters.  This is mostly why I wanted to add Travis, and our Kafka folks who maybe can answer this.  
-We could potentially end up returning multiple ACLs if we don't search on EXACT terms.
-We will need to compare the k8s configuration constantly to make sure that if there was a change to the YAML that it then gets picked up in k8s/crossplane, and goes and does the search.  When it realizes there's a discrepancy, it has to be a delete and a recreate.
- This plan doesn't work at all like topics, because you can grab those by their unique name.

2. Figure out a way to map one k8s managed ACL to one Kafka ACL
Pros:
- This would ideally work the best, since we have two unique IDs to map
Cons:
- May have to actually manage a map of some kind
- Could be massive
- Only possible if there is some other way to get ACLs by ID or reference them directly.

3. Don't do this, at all.
I don't think there's pros to this other than writing something bad, and the cons are quite long.  This might be an option if there's a change in libraries or tech that would help this.  We can revisit.

TL;DR:
We are using the same name for each ACL, K8s overwrites whatever is there (except it can't because there's no update option), so it ends up creating a new one and leaving orphaned ACLs
We're definitely concerned over performance if k8s has to constantly hit the describe ACL, but I want to verify that with Travis because it is purely an assumption.
We want to make sure that these providers are written in a way that they should work, and maintain that relationship between k8s managed resources and the Kafka brokers.
We're afraid that by describing the ACLs constantly, that we may accidentally clobber other ACLs in the process, when the K8s manifests change.  I can elaborate if need be.

Happy for any and all thoughts, but we're a little stuck.

[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:28]: @marshmallory I feel like 2nd solution could be possible.

AFAIU, there is no "id" or "name" of an ACL on the Kafka side.  So, to uniquely identify an ACL, we need to use all the fields in spec that build the ACL rule. What I am proposing is, let's use a custom external name (instead of using the name of the k8s resource) which is calculated using all fields like `User:Ed|ALLOW|ALTER_CONFIGS|TOPIC|marshmallory2|LITERAL|*` or even better as a `json` string like `{"PRINCIPAL":"User:Ed", "PERMISSION", "ALLOW", ...}`.

And use a flow like: 

image

[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:28]: Ah I forgot to add checking that acl really exist on Kafka side at the end of Observe flow, and go to create if not. This would re-create rule if deleted from Kafka.

[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:28]: Got it.  I understand this now.  Thanks for the diagram.

I think I have some questions, but I'll revisit with Ed and Jon and ask them as needed. Thanks :)

[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:28]: Digging into the code now.  

I now fully understand the flow of how the name is set, but I am unsure of how to solve the problem of the name being set by default.

I see that it hits the reconciler, and that's where it creates the struct for the managed object (or something along those lines):

```go
	r := managed.NewReconciler(mgr,
		resource.ManagedKind(v1alpha1.AccessControlListGroupVersionKind),
		managed.WithExternalConnecter(&connector{
			kube:         mgr.GetClient(),
			usage:        resource.NewProviderConfigUsageTracker(mgr.GetClient(), &apisv1alpha1.ProviderConfigUsage{}),
			newServiceFn: kafka.NewAdminClient}),
		managed.WithLogger(l.WithValues("controller", name)),
		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
```
Where NewReconciler is a struct with managed in it:
```go
	r := &Reconciler{
		client:       m.GetClient(),
		newManaged:   nm,
		pollInterval: defaultpollInterval,
		timeout:      reconcileTimeout,
		managed:      defaultMRManaged(m),
		external:     defaultMRExternal(),
		log:          logging.NewNopLogger(),
		record:       event.NewNopRecorder(),
	}
```
Where defaultMRManaged has the NewNameAsExternalName within:
```go
func defaultMRManaged(m manager.Manager) mrManaged {
	return mrManaged{
		ConnectionPublisher: NewAPISecretPublisher(m.GetClient(), m.GetScheme()),
		Finalizer:           resource.NewAPIFinalizer(m.GetClient(), managedFinalizerName),
		Initializer:         NewNameAsExternalName(m.GetClient()),
		ReferenceResolver:   NewAPISimpleReferenceResolver(m.GetClient()),
	}
}
```
And finally it resets the name:
```go
// NewNameAsExternalName returns a new NameAsExternalName.
func NewNameAsExternalName(c client.Client) *NameAsExternalName {
	return &NameAsExternalName{client: c}
}
```

Do I override the name after the fact, and consider the `Name=ExternalName` to be "unset" if it doesn't fit our new generator?

I was expecting it to be unset totally (or to be able to do that), but then I realized that crossplane programmatically sets the external-name via the code above.

I could override it, I assume, with the Get/Set methods available in the meta.go, but that means the original name or an "unset" name would still always be Name=ExternalName

Alternatively, if there's another way to override it, that would be great.

Overriding it in the annotation doesn't do much, because it is just pulling from the metadata name.

I have a few ideas, including adding an extra field, but I'd love to know your thoughts.

[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:28]: I should add, I did attempt to override it by directly setting a new external name via `SetExternalName()`, but it immediately gets overwritten and makes the name "marshmallory" again.

[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:28]: I think you should disable initializer as here:

[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:28]: Afk now, can check further if it doesn't help.

Btw, It is great that you've used this as an opportunity to dive into the crossplane runtime 🙌

[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:28]: Please continue your AFK!

Just typing this out for tomorrow :)

https://github.com/crossplane/provider-aws/blob/b3f7c2b86595f053bf9295cd05dfac5a71d1535a/pkg/controller/servicediscovery/httpnamespace/setup.go#L60

Are you referring to the `managed.WithInitializers()` function in the above link?

Our setup function:
```go
func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
	name := managed.ControllerName(v1alpha1.AccessControlListGroupKind)

	o := controller.Options{
		RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
	}

	r := managed.NewReconciler(mgr,
		resource.ManagedKind(v1alpha1.AccessControlListGroupVersionKind),
		managed.WithExternalConnecter(&connector{
			kube:         mgr.GetClient(),
			usage:        resource.NewProviderConfigUsageTracker(mgr.GetClient(), &apisv1alpha1.ProviderConfigUsage{}),
			newServiceFn: kafka.NewAdminClient}),
		managed.WithLogger(l.WithValues("controller", name)),
		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
		managed.WithInitializers())

	return ctrl.NewControllerManagedBy(mgr).
		Named(name).
		WithOptions(o).
		For(&v1alpha1.AccessControlList{}).
		Complete(r)
}
```

I definitely don't think I have the usage or customization correct, but I couldn't find any docs.

[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:28]: Yes @marshmallory looking correct 👍 

[It](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/reconciler.go#L556) will override the [default `Inializer` which contains NameAsExternalName](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/reconciler.go#L468) which would disable setting name as external name during [resource initialization](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/reconciler/managed/reconciler.go#L695).

**Comment by jograca** [comment]
Hope we have an MVP solution for Updates here @turkenh. We do a reconcile at Late Initialize. Would like your thoughts

File: internal/clients/kafka/acl/acl.go [hunk 9]
```diff
@@ -162,8 +162,16 @@ func Generate(name string, params *v1alpha1.AccessControlListParameters) *Access
 }
 
 // LateInitializeSpec fills empty ACL spec fields with the data retrieved from Kafka.
-func LateInitializeSpec() bool {
-	lateInitialized := true
+func LateInitializeSpec(in *v1alpha1.AccessControlListParameters, observed *AccessControlList) bool {
+	lateInitialized := false
```
Comment (turkenh) [inline internal/clients/kafka/acl/acl.go:166]: We late initialize optional parameters to reflect their defaults on the server side. 

With ACL resources spec, we don't have any optional parameters, so, I don't think we need to late-init anything. Hence, we can simply return `lateInialized=false` in ExternalObservation.

**Comment by turkenh** [comment]
> Hope we have an MVP solution for Updates here @turkenh. We do a reconcile at Late Initialize. Would like your thoughts

@jograca Any other questions than late init which I commented above?

File: internal/clients/kafka/acl/acl.go (hunk 9 above)
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:166]: Hasan, this was actually my suggestion and misunderstanding.

We have a situation where this workflow is broken:  
User applies a yaml -> ACL gets created -> External name get set (to the applied yaml) -> User attempts to update (by applying new YAML) -> External name != new yaml spec in k8s -> K8s thinks its an update/the manifest mismatches

My idea was just to use the LateInitialize to set the k8s configuration back to the actual state of the Kafka server.

What would be the correct way to do this?

Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:166]: I mean handling this mismatch in Observe and returning an error:

```
-> External name != new yaml spec in k8s -> Observe returns a proper error
```

Comment (turkenh) [reply internal/clients/kafka/acl/acl.go:166]: Yes, assuming you mean the annotation.

Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:166]: Oh, I see.  

Yes, we're handling that mismatch and returning the error, but we weren't sure how to get it to not update.  Also, I believe the YAML  in k8s needs to be the returned to the original for the delete, otherwise, it won't delete.  So we have to set it "back" to the original YAML somehow.  

I'll use the annotation.  Thanks :)

File: internal/controller/acl/acl.go [hunk 10]
```diff
@@ -51,6 +51,7 @@ const (
 	errGetCreds             = "cannot get credentials"
 	errListACL              = "cannot List ACLs"
 	errNewClient            = "cannot create new Service"
+	errUpdateNotSupported	= "upadtes are not supported"
```
Comment (elohmrow) [inline internal/controller/acl/acl.go:54]: s/upadtes/updates/

File: internal/controller/acl/acl.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+		return managed.ExternalObservation{}, errors.New(errNotAccessControlList)
+	}
+
+	// Check if the external name is set, to determine if ACL has been created or not
+	ext := meta.GetExternalName(cr)
+	if ext == "" {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	extname, err := acl.ConvertFromJSON(meta.GetExternalName(cr))
```
Comment (turkenh) [inline internal/controller/acl/acl.go:145]: error needs to be handled.

File: internal/controller/acl/acl.go [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+	compare := acl.CompareAcls(*extname, *acl.Generate(cr.Name, &cr.Spec.ForProvider))
+
+	if !compare {
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: false,
+		}, errors.Wrap(err, "Updating not allowed")
```
Comment (turkenh) [inline internal/controller/acl/acl.go:152]: including diff as part of error would be helpful to figure out what is not up to date.

File: internal/controller/acl/acl.go [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errListACL)
+	}
+
+	if ae == nil {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	cr.Status.AtProvider.ID = ae.Name
```
Comment (turkenh) [inline internal/controller/acl/acl.go:165]: Not sure if this provides meaningful information. We should drop `ID` from status otherwise.

File: internal/controller/acl/acl.go [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+	cr.Status.SetConditions(v1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        acl.IsUpToDate(&cr.Spec.ForProvider, ae),
```
Comment (turkenh) [inline internal/controller/acl/acl.go:170]: I think we no longer need to check if ACL is up to date at this point. We don't support updating and already errored above.

Comment (turkenh) [reply internal/controller/acl/acl.go:170]:
```suggestion
		ResourceUpToDate:        true,
```

File: internal/controller/acl/acl.go [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+
+	cr, ok := mg.(*v1alpha1.AccessControlList)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotAccessControlList)
+	}
+
+	generated := acl.Generate(cr.ObjectMeta.Name, &cr.Spec.ForProvider)
+	extname, err := acl.ConvertToJSON(generated)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.New("could not convert external name to JSON")
```
Comment (turkenh) [inline internal/controller/acl/acl.go:185]:
```suggestion
		return managed.ExternalCreation{}, errors.Wrap(err, "could not convert external name to JSON")
```

File: internal/controller/acl/acl.go [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+	}
+	if meta.GetExternalName(cr) == "" {
+		meta.SetExternalName(cr, extname)
+	}
+
+	return managed.ExternalCreation{ExternalNameAssigned: true}, acl.Create(ctx, c.kafkaClient, generated)
```
Comment (turkenh) [inline internal/controller/acl/acl.go:191]:
```suggestion
	if meta.GetExternalName(cr) == "" {
		meta.SetExternalName(cr, extname)
		return managed.ExternalCreation{ExternalNameAssigned: true}, acl.Create(ctx, c.kafkaClient, generated)
	}

	return managed.ExternalCreation{}, acl.Create(ctx, c.kafkaClient, generated)	
```

File: internal/clients/kafka/acl/acl.go [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+	b := kadm.ACLBuilder{}
+	ab := b.Allow(accessControlList.Principle).AllowHosts(accessControlList.Host).Operations(ao[0]).ResourcePatternType(rpt)
+
+	switch accessControlList.ResourceType {
+	case "Topic":
+		ab = ab.Topics(accessControlList.Name)
+	case "Group":
+		ab = ab.Groups(accessControlList.Name)
+	case "TransactionalID":
+		ab = ab.TransactionalIDs(accessControlList.Name)
```
[the remark this answers is not in the document]
Comment (jograca) [reply internal/clients/kafka/acl/acl.go]: We went through adding support for Cluster and Any, but found in both cases that it broke our Kafka clusters. We checked with our Kafka group and found that they were not supporting resource types outside of Topic and made a decision to remove it. Do you think we should add this in even if it might break the clusters for our users?

[the remark this answers is not in the document]
Comment (turkenh) [reply internal/clients/kafka/acl/acl.go]: It is interesting that try to add an ACL for Cluster type to break Kafka clusters. I would expect more graceful handling.

Considering this provider to be a more generic Kafka provider I think we should support it. In the worst case, I can think of dropping `Any` and `Cluster` from enum list for the field.

File: README.md [hunk 18]
```diff
@@ -84,34 +89,29 @@ parameters [here](https://github.com/bitnami/charts/tree/master/bitnami/kafka/#i
       sudo kubefwd svc -n kafka-cluster
       ```
 
-5. (optional) Install [kafka cli](https://github.com/birdayz/kaf).
+5. (optional) Install the [kafka cli](https://github.com/twmb/kcl).
```
Comment (twmb) [inline README.md:92]: For what it's worth, kcl is a bit out of date and needs some love.

File: internal/clients/kafka/acl/acl.go [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+
+	ao := []kadm.ACLOperation{o}
+
+	rpt, err := kmsg.ParseACLResourcePatternType(strings.ToLower(accessControlList.ResourcePatternTypeFilter))
+	if err != nil {
+		return nil, errors.Wrap(err, "did not return parsing of ACL pattern")
+	}
+
+	b := kadm.ACLBuilder{}
+	ab := b.Allow(accessControlList.ResourcePrinciple).AllowHosts(accessControlList.ResourceHost).Operations(ao[0]).ResourcePatternType(rpt)
```
Comment (stevendborrelli) [inline internal/clients/kafka/acl/acl.go:44]: Is there a case where  `ao[0]` doesn't exist?

File: apis/acl/v1alpha1/acl_types.go [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// AccessControlListParameters are the configurable fields of a AccessControlList.
+type AccessControlListParameters struct {
+	// +kubebuilder
+	ResourceName	string `json:"resourceName"`
```
[the remark this answers is not in the document]
Comment (jograca) [reply apis/acl/v1alpha1/acl_types.go]: Added descriptions and generated new CRDs

File: internal/clients/kafka/acl/acl.go (hunk 19 above)
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go]: Not that we've found, no, which is why there's no check here.

File: internal/clients/kafka/acl/acl.go [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+	case "Any":
+		ab = ab.AnyResource(accessControlList.ResourceName)
+	}
+
+	resp, err := cl.CreateACLs(ctx, ab)
+	if err != nil {
+		return err
+	}
+
+	a := resp[0].Principal
```
[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/clients/kafka/acl/acl.go:111]: Yes.

File: internal/clients/kafka/acl/acl.go (hunk 19 above)
Comment (jograca) [reply internal/clients/kafka/acl/acl.go]: We don't have a situation where ao[0] would not exist fortunately

**Review by stevendborrelli — APPROVED** [review]

File: internal/clients/kafka/acl/acl.go [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+	ResourceType              string `json:"ResourceType"`
+	ResourcePrincipal         string `json:"ResourcePrincipal"`
+	ResourceHost              string `json:"ResourceHost"`
+	ResourceOperation         string `json:"ResourceOperation"`
+	ResourcePermissionType    string `json:"ResourcePermissionType"`
+	ResourcePatternTypeFilter string `json:"ResourcePatternTypeFilter"`
+}
+
+// List lists all the ACLs in Kafka
+func List(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) (*AccessControlList, error) {
```
Comment (stevendborrelli) [inline internal/clients/kafka/acl/acl.go:29]: Linting is coming up with one more issue `cyclomatic complexity 11 of func `List` is high (> 10)`

Cyclomatic complexity is used as a metric for the complexity of tests required and we see it often in controllers. (see https://github.com/fzipp/gocyclo).

Maybe in future releases some of the logic can be factored out into functions. 

To fix, add the ` // nolint:gocyclo` comment to the function like so:

```
func List(ctx context.Context, cl *kadm.Client, accessControlList *AccessControlList) (*AccessControlList, error) { // nolint:gocyclo
```

**Review by stevendborrelli — APPROVED** [review]
I've approved it, but there is one linting error that comes up. (See my comment).

**Comment by elohmrow** [comment]
🎉

**Comment by turkenh** [comment]
🙌
